knimin/lib/vioscreen.py

```diff
@@ -0,0 +1,446 @@
+                             'Authorization': 'Bearer %s' % self.get_token()}
+        return self.get('https://api.viocare.com/%s/users' % self._key,
+                         headers=self._headers)
+
+    def get(self, url, retries=5, **kwargs):
+        for i in range(retries):
+            req = self._session.get(url, **kwargs)
+            if req.status_code != 200:  # HTTP status code, 200 is all good
+                data = req.json()
+
+                # if we did not get a HTTP status code 200, than guess that the
+                # API token is no longer valid so get a new one and retry
+                if 'Code' in data and data['Code'] == 1016:
+                    self._headers['token'] = self.get_token()
+                else:
+                    print(self._session.get, url, kwargs)
+                    raise ValueError("Unable to make this query work")
+            else:
+                return req.json()
+        raise ValueError("Unable to make this query work")
+
+    def post(self, url, retries=5, **kwargs):
+        for i in range(retries):
+            req = self._session.post(url, **kwargs)
+            if req.status_code != 200:  # HTTP status code, 200 is all good
+                data = req.json()
+
+                # if we did not get a HTTP status code 200, than guess that the
+                # API token is no longer valid so get a new one and retry
+                if 'Code' in data and data['Code'] == 1016:
+                    self._headers['token'] = self.get_token()
+                else:
+                    print(self._session.post, url, kwargs)
+                    raise ValueError("Unable to make this query work")
+            else:
+                return req.json()
+        raise ValueError("Unable to make this query work")
+
+    def tidyfy(self, username, payload):
+        """Restructures data so that 'survey_id' is associated with each row
+
+        Parameters
+        ----------
+        username: str
+            Survey ID that is being inserted each of the rows
+        payload: list of dict
+            The data that is getting associated with the survey_id
+
+        Return
+        ------
+        list of dict
+            The newly formatted data containing the username in each row
+        """
+        dat = []
+        for entry in payload:
+            entry['survey_id'] = username
+            dat.append(entry)
+        return dat
+
+    def get_session_data(self, session_id, endpoint):
+        """Pulls data from the vioscreen API based on 
+           a specific session ID and session type(ex. 'foodcomponents')
+
+        Parameters
+        ----------
+        session_id: str
+            Session ID that data is being requested for
+        endpoint: str
+            Name of the session type that data is being requested for
+
+        Return
+        ------
+        dict:
+            Food frequency questionnaire data
+        """
+        if not self._headers:
```

Comment:
It looks like this if-statement is replicated logic, should it be centralized into a boilerplate method? One strategy here would be to make an idempotent function, something like:

```python
class VioscreenHandler(object):
    ...
    def _assure_headers(self):
        if not self._headers:
            self._headers = {'Accept': 'application/json',
                             'Authorization': 'Bearer %s' % self.get_token()}
```

...and then in methods which need the headers, they could always just have the boilerplate call:

```python
class VioscreenHandler(object):
    ...
    def get_session_data(...):
        """ ... """
        self._assure_headers()
        ...
```